tools/codegen/dest/lazy_ir.py

```diff
@@ -0,0 +1,138 @@
+            if str(t.elem) == 'Tensor':
+                types.append(NamedCType(binds, BaseCType(valueListT)))
+                value_types.append(NamedCType(binds, BaseCType(valueListT)))
+            elif str(t.elem) == 'Tensor?':
+                types.append(NamedCType(binds, ConstRefCType(ListCType(OptionalCType(BaseCType(valueT))))))
+                value_types.append(NamedCType(binds, ConstRefCType(ListCType(OptionalCType(BaseCType(valueT))))))
+            else:
+                types.append(cpp.argumenttype_type(t, mutable=mutable, binds=binds))
+                scalar_types.append(cpp.argumenttype_type(t, mutable=mutable, binds=binds))
+        else:
+            raise AssertionError(f"unrecognized type {repr(t)}")
+    return types, value_types, scalar_types
+
+@dataclass(frozen=True)
+class LazyIR:
+    backend_index: BackendIndex
+
+    target: Union[
+        Literal[Target.DEFINITION],
+        Literal[Target.DECLARATION],
+    ]
+
+    # TODO(whc) probably use selector instead of building a separate index for ops to codegen
+    # Selector object to determine which operators to generate
+    # registration code for.
+    selector: SelectiveBuilder
+
+    # The namespace that the kernels are written in. This is just `at::native` for in-tree kernels.
+    cpp_namespace: str
+
+    @method_with_native_function
+    def __call__(self, f: Union[NativeFunctionsGroup, NativeFunction]) -> List[str]:
+        func = f.functional.func if isinstance(f, NativeFunctionsGroup) else f.func
+        if func.name in self.backend_index.index:
+            return self.gen(f)
+        else:
+            return []
+
+    def gen(self, f: Union[NativeFunctionsGroup, NativeFunction]) -> List[str]:
+        # for now, we just want one IR class decl and soon after also the method defs
+        # and we use the functional version not out/inplace.
+        func = f.functional.func if isinstance(f, NativeFunctionsGroup) else f.func
+        class_name = str(func.name).lower().capitalize()
+
+        all_types, value_types, scalar_types = process_ir_types(func)
+        node_ctor_args = ", ".join([f"{i.cpp_type()} {i.name}" for i in all_types])
+        scalar_initializers = ",\n        ".join([f"{t.name}_({t.name})" for t in scalar_types])
+        scalar_decls = ";\n  ".join([f"{t.cpp_type()} {t.name}_" for t in scalar_types])
+        scalar_hashes = ", ".join([f.name for f in scalar_types])
+        return [f"""\
+class {class_name} : public LazyNodeBase {{
+ public:
+  {class_name}({node_ctor_args})
+      : LazyNodeBase(ir::OpKind(at::aten::{func.name.name}),
+              {{{", ".join([f"{i.name}" for i in value_types])}}},
+              /*num_outputs=*/{len(func.returns)},
+              lazy_tensors::util::MHash({scalar_hashes})),
+        {scalar_initializers}
+  {{}}
+
+  std::string ToString() const override;
+
+  NodePtr Clone(OpList operands) const override;
+
+ private:
+  std::vector<c10::Optional<ir::Value>> inputs;
```

Comment:
It's kind of weird that all the scalars are unboxed, but the inputs are stashed into a vector. Maybe easier to just have them unboxed too.

Reply:
Yea, I looked into this a bit and it's because the base class is tracking usages of Values for liveness purposes.  But this doesn't mean they have to be in a vector, it just means they need special handling, which we could generate another way.